In [1]:
import pandas as pd
import json
import ast
import os
from transformers import MarkupLMProcessor

# SKIP THE BELOW CODE UNTIL THE CELL THAT SAYS "SKIP TO HERE.." GO TO BOTTOM OF NOTEBOOK!

In [1]:
########################################################################################################################

In [2]:
base_path = r'C:\Users\islam\Desktop\2023 Research\contracts/'
preds_path = 'preds/'
preds_jsons_path = 'preds/preds_jsons/'
contract_num = '128'
full_contract_csv_path = fr'C:\Users\islam\Desktop\2023 Research\contracts\tagged\test/contract_{contract_num}_tagged.csv'

inpath = f'{base_path}{preds_path}contract_{contract_num}_preds_tokenized.csv'
outpath = f'{base_path}{preds_jsons_path}contract_{contract_num}_preds.json'

start_row_idx = 1100 ## article ii the merger node
stop_row_idx = 1174 ## page 21 node

In [3]:
df = pd.read_csv(full_contract_csv_path).iloc[start_row_idx:stop_row_idx].copy().dropna()
df

Unnamed: 0                                             xpaths  \
1100        1100  /html/body/document/type/sequence/filename/des...   
1101        1101  /html/body/document/type/sequence/filename/des...   
1102        1102  /html/body/document/type/sequence/filename/des...   
1103        1103  /html/body/document/type/sequence/filename/des...   
1108        1108  /html/body/document/type/sequence/filename/des...   
1109        1109  /html/body/document/type/sequence/filename/des...   
1116        1116  /html/body/document/type/sequence/filename/des...   
1117        1117  /html/body/document/type/sequence/filename/des...   
1128        1128  /html/body/document/type/sequence/filename/des...   
1129        1129  /html/body/document/type/sequence/filename/des...   
1132        1132  /html/body/document/type/sequence/filename/des...   
1133        1133  /html/body/document/type/sequence/filename/des...   
1134        1134  /html/body/document/type/sequence/filename/des...   
1136        1136  /html/body/document/type/sequence/filename/des...   
1137        1137  /html/body/document/type/sequence/filename/des...   
1147        1147  /html/body/document/type/sequence/filename/des...   
1148        1148  /html/body/document/type/sequence/filename/des...   
1153        1153  /html/body/document/type/sequence/filename/des...   
1154        1154  /html/body/document/type/sequence/filename/des...   
1156        1156  /html/body/document/type/sequence/filename/des...   
1157        1157  /html/body/document/type/sequence/filename/des...   
1160        1160  /html/body/document/type/sequence/filename/des...   
1161        1161  /html/body/document/type/sequence/filename/des...   
1164        1164  /html/body/document/type/sequence/filename/des...   
1165        1165  /html/body/document/type/sequence/filename/des...   
1167        1167  /html/body/document/type/sequence/filename/des...   
1168        1168  /html/body/document/type/sequence/filename/des...   
1173        1173  /html/body/document/type/sequence/filename/des...   

                                         text  all_nodes_ordering  \
1100                               ARTICLE II                1100   
1101                               THE MERGER                1101   
1102                                      2.1                1102   
1103                               The Merger                1103   
1108                                      2.2                1108   
1109                       The Effective Time                1109   
1116                                      2.3                1116   
1117                              The Closing                1117   
1128                                      2.4                1128   
1129                     Effect of the Merger                1129   
1132                                       20                1132   
1133                                      2.5                1133   
1134  Certificate of Incorporation and Bylaws                1134   
1136                                      (a)                1136   
1137             Certificate of Incorporation                1137   
1147                                      (b)                1147   
1148                                   Bylaws                1148   
1153                                      2.6                1153   
1154                   Directors and Officers                1154   
1156                                      (a)                1156   
1157                                Directors                1157   
1160                                      (b)                1160   
1161                                 Officers                1161   
1164                                      2.7                1164   
1165                  Effect on Capital Stock                1165   
1167                                      (a)                1167   
1168                            Capital Stock                1168   
1173          

In [4]:
processor = MarkupLMProcessor.from_pretrained("microsoft/markuplm-base-finetuned-websrc")
processor.parse_html = False

In [5]:
nodes = df.highlighted_segmented_text.to_list()
xpaths = df.highlighted_xpaths.to_list()
tokens = []
for i in range(len(df)):
  encoding = processor(nodes=nodes[i:i+1], xpaths=xpaths[i:i+1], return_tensors="pt", truncation=False)
  tokens.append([processor.decode(x) for x in encoding['input_ids'][0]][1:-1])

In [6]:
df['tokens'] = tokens
df

Unnamed: 0                                             xpaths  \
1100        1100  /html/body/document/type/sequence/filename/des...   
1101        1101  /html/body/document/type/sequence/filename/des...   
1102        1102  /html/body/document/type/sequence/filename/des...   
1103        1103  /html/body/document/type/sequence/filename/des...   
1108        1108  /html/body/document/type/sequence/filename/des...   
1109        1109  /html/body/document/type/sequence/filename/des...   
1116        1116  /html/body/document/type/sequence/filename/des...   
1117        1117  /html/body/document/type/sequence/filename/des...   
1128        1128  /html/body/document/type/sequence/filename/des...   
1129        1129  /html/body/document/type/sequence/filename/des...   
1132        1132  /html/body/document/type/sequence/filename/des...   
1133        1133  /html/body/document/type/sequence/filename/des...   
1134        1134  /html/body/document/type/sequence/filename/des...   
1136        1136  /html/body/document/type/sequence/filename/des...   
1137        1137  /html/body/document/type/sequence/filename/des...   
1147        1147  /html/body/document/type/sequence/filename/des...   
1148        1148  /html/body/document/type/sequence/filename/des...   
1153        1153  /html/body/document/type/sequence/filename/des...   
1154        1154  /html/body/document/type/sequence/filename/des...   
1156        1156  /html/body/document/type/sequence/filename/des...   
1157        1157  /html/body/document/type/sequence/filename/des...   
1160        1160  /html/body/document/type/sequence/filename/des...   
1161        1161  /html/body/document/type/sequence/filename/des...   
1164        1164  /html/body/document/type/sequence/filename/des...   
1165        1165  /html/body/document/type/sequence/filename/des...   
1167        1167  /html/body/document/type/sequence/filename/des...   
1168        1168  /html/body/document/type/sequence/filename/des...   
1173        1173  /html/body/document/type/sequence/filename/des...   

                                         text  all_nodes_ordering  \
1100                               ARTICLE II                1100   
1101                               THE MERGER                1101   
1102                                      2.1                1102   
1103                               The Merger                1103   
1108                                      2.2                1108   
1109                       The Effective Time                1109   
1116                                      2.3                1116   
1117                              The Closing                1117   
1128                                      2.4                1128   
1129                     Effect of the Merger                1129   
1132                                       20                1132   
1133                                      2.5                1133   
1134  Certificate of Incorporation and Bylaws                1134   
1136                                      (a)                1136   
1137             Certificate of Incorporation                1137   
1147                                      (b)                1147   
1148                                   Bylaws                1148   
1153                                      2.6                1153   
1154                   Directors and Officers                1154   
1156                                      (a)                1156   
1157                                Directors                1157   
1160                                      (b)                1160   
1161                                 Officers                1161   
1164                                      2.7                1164   
1165                  Effect on Capital Stock                1165   
1167                                      (a)                1167   
1168                            Capital Stock                1168   
1173          

In [7]:
rename_cols = {
    'segment_number_from_idx':'seg_num',
    'highlighted_segmented_text':'h_text',
}
cols_to_keep = [
    'top',
    'left',
    'width',
    'height',
    'tagged_sequence',
    'seg_num',
    'preds',
    'tokens',
    'h_text'
]

In [8]:
df['preds'] = -1 # manually fill in with your preds
output = df.rename(columns=rename_cols)
output = output[cols_to_keep].copy(deep=True)
output['seg_num'] = output['seg_num'].astype(int)

In [9]:

#output = output.drop('c', axis=1)
#output['tokens'] = output['tokens'].apply(ast.literal_eval)

In [10]:
output

top      left     width    height tagged_sequence  seg_num  preds  \
1100  0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
1101  0.265848  0.467906  0.064189  0.000333            s_st       35     -1   
1102  0.266536  0.039298  0.011694  0.000333           s_ssn       36     -1   
1103  0.266536  0.081606  0.043766  0.000333           s_sst       37     -1   
1108  0.267568  0.039298  0.014364  0.000333           s_ssn       38     -1   
1109  0.267568  0.077692  0.070285  0.000333           s_sst       39     -1   
1116  0.268944  0.039298  0.011694  0.000333           s_ssn       40     -1   
1117  0.268944  0.071985  0.045768  0.000333           s_sst       41     -1   
1128  0.270664  0.039298  0.011694  0.000333           s_ssn       42     -1   
1129  0.270664  0.073844  0.074808  0.000333           s_sst       43     -1   
1132  0.272336  0.496255  0.007489  0.000266             s_n       44     -1   
1133  0.273044  0.039298  0.011694  0.000333           s_ssn       45     -1   
1134  0.273044  0.074380  0.150181  0.000333           s_sst       46     -1   
1136  0.273732  0.072982  0.012780  0.000333          s_sssn       47     -1   
1137  0.273732  0.104928  0.104561  0.000333          s_ssst       48     -1   
1147  0.275108  0.072982  0.010910  0.000333          s_sssn       49     -1   
1148  0.275108  0.106025  0.027023  0.000333          s_ssst       50     -1   
1153  0.276137  0.039298  0.011694  0.000333           s_ssn       51     -1   
1154  0.276137  0.074380  0.084359  0.000333           s_sst       52     -1   
1156  0.276825  0.072982  0.010384  0.000333          s_sssn       53     -1   
1157  0.276825  0.109238  0.035510  0.000333          s_ssst       54     -1   
1160  0.277860  0.072982  0.013372  0.000333          s_sssn       55     -1   
1161  0.277860  0.106047  0.030137  0.000333          s_ssst       56     -1   
1164  0.278889  0.039298  0.011694  0.000333           s_ssn       57     -1   
1165  0.278889  0.074380  0.086513  0.000333           s_sst       58     -1   
1167  0.279577  0.072982  0.010384  0.000333          s_sssn       59     -1   
1168  0.279577  0.109644  0.050954  0.000333          s_ssst       60     -1   
1173  0.281939  0.496255  0.007489  0.000266             s_n       61     -1   

                                                 tokens  \
1100                                   [ART, ICLE,  II]   
1101                                   [THE,  MER, GER]   
1102                                          [2, ., 1]   
1103                                   [The,  Mer, ger]   
1108                                          [2, ., 2]   
1109                           [The,  Effective,  Time]   
1116                                          [2, ., 3]   
1117                                    [The,  Closing]   
1128                                          [2, ., 4]   
1129                     [Effect,  of,  the,  Mer, ger]   
1132                                               [20]   
1133                                          [2, ., 5]   
1134  [Cert, ificate,  of,  Inc, orpor, ation,  and,...   
1136                                          [(, a, )]   
1137           [Cert, ificate,  of,  Inc, orpor, ation]   
1147                                          [(, b, )]   
1148                                       [B, yl, aws]   
1153                                          [2, ., 6]   
1154                     [Direct, ors,  and,  Officers]   
1156                                          [(, a, )]   
1157                                      [Direct, ors]   
1160                                          [(, b, )]   
1161                                         [Officers]   
1164                                          [2, ., 7]   
1165                    [Effect,  on,  Capital,  Stock]   
1167                                          [(, a, )]   
1168                                  [Capital,  Stock]   
1173                                  

In [11]:
output = output.explode(column=['tokens']).reset_index(drop=True)

In [12]:
output

top      left     width    height tagged_sequence  seg_num  preds  \
0   0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
1   0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
2   0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
3   0.265848  0.467906  0.064189  0.000333            s_st       35     -1   
4   0.265848  0.467906  0.064189  0.000333            s_st       35     -1   
..       ...       ...       ...       ...             ...      ...    ...   
84  0.279577  0.072982  0.010384  0.000333          s_sssn       59     -1   
85  0.279577  0.072982  0.010384  0.000333          s_sssn       59     -1   
86  0.279577  0.109644  0.050954  0.000333          s_ssst       60     -1   
87  0.279577  0.109644  0.050954  0.000333          s_ssst       60     -1   
88  0.281939  0.496255  0.007489  0.000266             s_n       61     -1   

     tokens         h_text  
0       ART     ARTICLE II  
1      ICLE     ARTICLE II  
2        II     ARTICLE II  
3       THE     THE MERGER  
4       MER     THE MERGER  
..      ...            ...  
84        a            (a)  
85        )            (a)  
86  Capital  Capital Stock  
87    Stock  Capital Stock  
88       21             21  

[89 rows x 9 columns]

In [13]:
output['t_len'] = output['tokens'].apply(len)
output['len'] = output['h_text'].apply(len)

Let's confirm the sum of the parts is the whole

In [14]:
output.groupby('seg_num').agg({'t_len':'sum','len':'first'})

t_len  len
seg_num            
34          10   10
35          10   10
36           3    3
37          10   10
38           3    3
39          18   18
40           3    3
41          11   11
42           3    3
43          20   20
44           2    2
45           3    3
46          39   39
47           3    3
48          28   28
49           3    3
50           6    6
51           3    3
52          22   22
53           3    3
54           9    9
55           3    3
56           8    8
57           3    3
58          23   23
59           3    3
60          13   13
61           2    2

Confirmed

In [15]:
output['pct_width'] = output['t_len'] / output['len']
output['width_added'] = output.width * output.pct_width

In [16]:
output['start'] = -1
output

top      left     width    height tagged_sequence  seg_num  preds  \
0   0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
1   0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
2   0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
3   0.265848  0.467906  0.064189  0.000333            s_st       35     -1   
4   0.265848  0.467906  0.064189  0.000333            s_st       35     -1   
..       ...       ...       ...       ...             ...      ...    ...   
84  0.279577  0.072982  0.010384  0.000333          s_sssn       59     -1   
85  0.279577  0.072982  0.010384  0.000333          s_sssn       59     -1   
86  0.279577  0.109644  0.050954  0.000333          s_ssst       60     -1   
87  0.279577  0.109644  0.050954  0.000333          s_ssst       60     -1   
88  0.281939  0.496255  0.007489  0.000266             s_n       61     -1   

     tokens         h_text  t_len  len  pct_width  width_added  start  
0       ART     ARTICLE II      3   10   0.300000     0.015576     -1  
1      ICLE     ARTICLE II      4   10   0.400000     0.020768     -1  
2        II     ARTICLE II      3   10   0.300000     0.015576     -1  
3       THE     THE MERGER      3   10   0.300000     0.019257     -1  
4       MER     THE MERGER      4   10   0.400000     0.025675     -1  
..      ...            ...    ...  ...        ...          ...    ...  
84        a            (a)      1    3   0.333333     0.003461     -1  
85        )            (a)      1    3   0.333333     0.003461     -1  
86  Capital  Capital Stock      7   13   0.538462     0.027437     -1  
87    Stock  Capital Stock      6   13   0.461538     0.023517     -1  
88       21             21      2    2   1.000000     0.007489     -1  

[89 rows x 14 columns]

In [17]:
seg_num = -1
for i, row in output.iterrows():
    
    # new segment
    if row.seg_num != seg_num:
        seg_num = row.seg_num
        output.at[i, 'start'] = row.left
    else:
        output.at[i, 'start'] = prev_end
    prev_end = output.at[i, 'start'] + row.width_added

In [18]:
output

top      left     width    height tagged_sequence  seg_num  preds  \
0   0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
1   0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
2   0.265504  0.474041  0.051919  0.000333            s_sn       34     -1   
3   0.265848  0.467906  0.064189  0.000333            s_st       35     -1   
4   0.265848  0.467906  0.064189  0.000333            s_st       35     -1   
..       ...       ...       ...       ...             ...      ...    ...   
84  0.279577  0.072982  0.010384  0.000333          s_sssn       59     -1   
85  0.279577  0.072982  0.010384  0.000333          s_sssn       59     -1   
86  0.279577  0.109644  0.050954  0.000333          s_ssst       60     -1   
87  0.279577  0.109644  0.050954  0.000333          s_ssst       60     -1   
88  0.281939  0.496255  0.007489  0.000266             s_n       61     -1   

     tokens         h_text  t_len  len  pct_width  width_added     start  
0       ART     ARTICLE II      3   10   0.300000     0.015576  0.474041  
1      ICLE     ARTICLE II      4   10   0.400000     0.020768  0.489616  
2        II     ARTICLE II      3   10   0.300000     0.015576  0.510384  
3       THE     THE MERGER      3   10   0.300000     0.019257  0.467906  
4       MER     THE MERGER      4   10   0.400000     0.025675  0.487163  
..      ...            ...    ...  ...        ...          ...       ...  
84        a            (a)      1    3   0.333333     0.003461  0.076444  
85        )            (a)      1    3   0.333333     0.003461  0.079905  
86  Capital  Capital Stock      7   13   0.538462     0.027437  0.109644  
87    Stock  Capital Stock      6   13   0.461538     0.023517  0.137080  
88       21             21      2    2   1.000000     0.007489  0.496255  

[89 rows x 14 columns]

In [21]:
output.to_csv(inpath)

In [ ]:
############################################################################################################################

# SKIP TO HERE TO AVOID MAKING THE ORIGINAL FILE AND GO STRAIGHT TO EDITING!

## Please git pull the contracts repo

## Now manually add the preds to the file with path `inpath`. We are editing contract 128


There should be a column called preds in that file path so edit it and then add the preds.

....


Now that you're done editing that file, we will now just read in and output the finalized json file. Please only do this AFTER you have manually edited the file for the predictions for each token

In [5]:
df = pd.read_csv(inpath)

FileNotFoundError: [Errno 2] No such file or directory: 'C:/Users/islam/Desktop/2023 Research/contracts/preds/contract_128_preds_tokenized.csv'

In [6]:
contract_num = '128'

## Replace with your path to the contracts repo
inpath = f'/Users/sukritrao/Documents/NYU/Coursework/Spring2023/Independent-Study/project/markuplm_playground/labeling_old/contracts/preds/contract_128_preds_tokenized.csv'
outpath = f'/Users/sukritrao/Documents/NYU/Coursework/Spring2023/Independent-Study/project/markuplm_playground/labeling_old/contracts/preds/preds_jsons/contract_{contract_num}_preds.json'

In [8]:
df = pd.read_csv(inpath)
json_cols = [
    'tagged_sequence',
    'seg_num',
    'preds',
    'top',
    'height',
    'start',
    'width_added'    
]
df[json_cols].to_json(outpath, orient='records')

KeyError: "['tagged_sequence', 'seg_num', 'top', 'height', 'start', 'width_added'] not in index"

Now please push the contract to github

it should be in `contracts/preds_jsons/` folder

Add in the visualizer code found in https://github.com/pranabislam/contract_labeler/blob/main/pred_visualizer_at_token_level.js into the javascript chrome add on and then go to contract 128. Look on page 20 - 21, article II through section 2.7

Then use the normal labeler code and screenshot the same contract sections
https://github.com/pranabislam/contract_labeler/blob/main/simple_visualizer.js